ishop/views.py

```diff
@@ -1,3 +1,197 @@
-from django.shortcuts import render
+from django.contrib.auth import login
+from django.contrib.auth.mixins import LoginRequiredMixin, UserPassesTestMixin
+from django.contrib.auth.views import LoginView, LogoutView
+from django.shortcuts import redirect
+from django.utils import timezone
+from django.views import View
+from django.views.generic import ListView, CreateView, UpdateView
+from django.contrib import messages
+from django.db import transaction
 
-# Create your views here.
+from ishop.forms import CustomUserCreationForm
+from ishop.models import Good, ShopUser, Purchase, Refund
+
+
+class SuperUserRequiredMixin(LoginRequiredMixin, UserPassesTestMixin):
+    def test_func(self):
+        return self.request.user.is_superuser
+
+
+class GoodsListView(ListView):
+    paginate_by = 10
+    http_method_names = ['post', 'get']
+    template_name = 'goods_list.html'
+    queryset = Good.objects.filter(in_stock__gt=0)
+    extra_context = {'title': 'Online shop'}
+
+
+class PurchaseView(View):
+    http_method_names = ['post', ]
+
+    def post(self, request, *args, **kwargs):
+        if not request.user.is_authenticated:
+            msg = "Only logged users can buy"
+            messages.warning(self.request, msg)
+            return redirect('login')
+
+        user = self.request.user
+        good = Good.objects.get(id=self.request.POST['pk'])
+        quantity = int(self.request.POST['quantity'])
+        price = good.price
+        in_stock = good.in_stock
+        wallet = user.wallet
+        amount = quantity * price
+
+        if wallet < amount:
+            msg = "You don't have enough money for this purchase"
+            messages.error(self.request, msg)
+            return redirect('goods')
+
+        if in_stock < quantity:
+            msg = "We don't have enough goods in stock"
+            messages.warning(self.request, msg)
+            return redirect('goods')
+
+        user.wallet -= amount
+        good.in_stock -= quantity
+        purchase = Purchase(customer=user, good=good, quantity=quantity, price=price)
+
+        with transaction.atomic():
+            user.save()
+            good.save()
+            purchase.save()
+
+        msg = "Your purchase is done"
+        messages.success(self.request, msg)
+
+        return redirect('goods')
+
+
+class PurchaseRefundView(View):
+    http_method_names = ['post', ]
+    interval_to_refund = 3  # in minutes
+
+    @staticmethod
+    def get_time_to_refund():
+        interval = PurchaseRefundView.interval_to_refund
+        return timezone.now() - timezone.timedelta(minutes=interval)
+
+    def post(self, request, *args, **kwargs):
+        purchase = Purchase.objects.get(pk=self.request.POST['pk'])
```

Comment:
Not sure about this solution, totally possible but maybe create view will be more convenient. In this case absolutely up to you

Reply:
CreateView requires creating a form, but my small forms are already created and filled on the ListView-pages and my goal is only to handle POST.
The same thing with creating Purchase, Refund, approving/declining Refunds by admin
I think better way use the usual View, in this case: we follow the principle of single responsibility and at the same time don't make User fill out any extra forms and press any additional buttons